pkg/autoscaler/statforwarder/forwarder.go

```diff
@@ -253,15 +274,80 @@ func (f *Forwarder) createOrUpdateEndpoints(ns, n string) error {
 	return nil
 }
 
-func (f *Forwarder) getHolder(name string) (string, bool) {
-	f.leaseLock.RLock()
-	defer f.leaseLock.RUnlock()
-	result, ok := f.leaseHolders[name]
+func (f *Forwarder) getProcessor(bkt string) (*bucketProcessor, bool) {
+	f.processorsLock.RLock()
+	defer f.processorsLock.RUnlock()
+	result, ok := f.processors[bkt]
 	return result, ok
 }
 
-func (f *Forwarder) setHolder(name, holder string) {
-	f.leaseLock.Lock()
-	defer f.leaseLock.Unlock()
-	f.leaseHolders[name] = holder
+func (f *Forwarder) setProcessor(bkt string, p *bucketProcessor) {
+	f.processorsLock.Lock()
+	defer f.processorsLock.Unlock()
+	f.processors[bkt] = p
+}
+
+func (f *Forwarder) createProcessor(bkt, holder string) *bucketProcessor {
+	if holder == f.selfIP {
+		return &bucketProcessor{
+			holder: holder,
+			proc: func(sm asmetrics.StatMessage) {
+				l := f.logger.With(zap.String("revision", sm.Key.String()))
+				l.Debug("Accept stat as owner of bucket ", bkt)
+				if f.accept != nil {
+					f.accept(sm)
+				}
+			},
+		}
+	}
+
+	// TODO(9235): Currently we use IP directly. This won't work if there
+	// is mesh. Need to fall back to connection via Service.
+	dns := fmt.Sprintf("ws://%s:%d", holder, autoscalerPort)
+	f.logger.Info("Connecting to Autoscaler bucket at ", dns)
+	c := websocket.NewDurableSendingConnection(dns, f.logger)
+	return &bucketProcessor{
+		holder: holder,
+		conn:   c,
+		proc: func(sm asmetrics.StatMessage) {
+			l := f.logger.With(zap.String("revision", sm.Key.String()))
+			l.Debugf("Forward stat of bucket %s to the holder %s", bkt, holder)
+			wsms := asmetrics.ToWireStatMessages([]asmetrics.StatMessage{sm})
+			b, err := wsms.Marshal()
+			if err != nil {
+				l.Errorw("Error while marshaling stats", zap.Error(err))
+				return
+			}
+
+			if err := c.SendRaw(gorillawebsocket.BinaryMessage, b); err != nil {
+				l.Errorw("Error while sending stats", zap.Error(err))
+			}
+		},
+	}
+}
+
+// Process calls Forwarder.accept if the pod where this Forwarder running is the owner
+// of the given StatMessage. Otherwise it forwards the given StatMessage to the right
+// owner pod. If it can't find the owner, it drops the StatMessage.
+func (f *Forwarder) Process(sm asmetrics.StatMessage) {
+	rev := sm.Key.String()
+	l := f.logger.With(zap.String("revision", rev))
+	bkt := f.bs.Owner(rev)
+
+	p, ok := f.getProcessor(bkt)
+	if !ok {
+		l.Warnf("Can't find the owner for Rev %s. Dropping its stat.", rev)
```

Comment:
Just to ask the question: in what cases can this happen? How common are the cases where this occurs and is it OK to drop the stat message in them (presumably meaning we might miss a scale-from-zero event)?

Reply:
In theory it happens when a Lease for a revision hasn't been acquired while there's traffic for this revision. I don't have actual data for how common it is. I think it could happen during serving upgrade, especially for the first time we introduce the leases. Instead of dropping, maybe we should cache and retry?